### DISCUSSION TITLE
[Graphcache] Equivalent to ApolloClient's writeQuery / writeFragment?

### DISCUSSION THREAD

**Discussion by ecklf** [urql-graphql/urql#1059] [Q&A] [answered]
I am considering migrating from Apollo Client to urql, but I am unsure of how I would solve writing to Graphcache without executing a Mutation. I want to optimize my requests to the API. Thus it doesn't feel right to me in some cases.

Any solutions to achieve something similar in the urql way?

**Comment by JoviDeCroock** [comment, marked as answer]
Hey,

GraphCache doesn't allow this with a reason, we have structured this cache to rely on the backend as a source of truth. This means that if we alter the cache outside of a mutation that we'd be straying away from our source of truth.
Doing so can lead to indeterministic behavior which is something we're aiming to avoid, [this](https://www.apollographql.com/docs/react/api/cache/InMemoryCache/#writequery) seems to indicate that the alteration you perform never exits the cache (it never updates the server-data) which results in a discrepancy for the data.

**Reply by ecklf** [reply]
@JoviDeCroock fair point

> this seems to indicate that the alteration you perform never exits the cache

Personally, I think it is more about being able to control _when_ your cached data exits the cache. I assume your proposed solution would be to use a state management solution to handle my modifications, but I like the idea of being able to modify my GraphQL data in one place.

It could still be useful to expose such functionality with _dangerously_ prefixed methods. "Dangerous" as in cached data being lost if you don't deal with handling the discrepancy you've created.

I can totally understand your decision. I just wanted to share my thoughts on this.

Thanks for the quick response! I appreciate it.

**Comment by Blankeos** [comment]
Any suggested workarounds for this in 2025? I have a really specific usecase where I want to use `useQuery` but I have to subscribe to a separate event (not in GraphQL) to update the state in `useQuery`. Is this possible with `cacheExchange`?
